nengo/neurons.py

```diff
@@ -727,7 +832,147 @@ def step_math(self, dt, J, spiked, voltage, recovery):
         recovery[spiked > 0] = recovery[spiked > 0] + self.reset_recovery
 
 
-class NeuronTypeParam(Parameter):
-    def coerce(self, instance, neurons):
-        self.check_type(instance, neurons, NeuronType)
-        return super().coerce(instance, neurons)
+class _Spiking(NeuronType):
+    """Abstract base class for neuron models that turn rate models into spiking ones"""
+
+    base_type = NeuronTypeParam("base_type")
+    amplitude = NumberParam("amplitude", low=0, low_open=True)
+
+    def __init__(self, base_type, amplitude=1.0):
+        super().__init__()
+
+        self.base_type = base_type
+        if "rates" not in base_type.probeable:
+            raise ValidationError(
+                "Must be a rate neuron type", attr="base_type", obj=self
+            )
+
+        self.negative = base_type.negative
+        self.amplitude = amplitude
+
+    @property
+    def probeable(self):
+        base = list(self.base_type.probeable)
+        base[base.index("rates")] = "spikes"
+        return tuple(base)
+
+    def gain_bias(self, max_rates, intercepts):
+        return self.base_type.gain_bias(max_rates, intercepts)
+
+    def max_rates_intercepts(self, gain, bias):
+        return self.base_type.max_rates_intercepts(gain, bias)
+
+    def rates(self, x, gain, bias):
+        return self.base_type.rates(x, gain, bias)
+
+
+class RegularSpiking(_Spiking):
+    """Turn a rate neuron model into a spiking one with regular inter-spike intervals.
+
+    Spikes at regular intervals based on the rates of the base neuron type. [1]_
+
+    Parameters
+    ----------
+    base_type : NeuronType
+        A rate-based neuron type to convert to a regularly spiking neuron.
+    amplitude : float
+        Scaling factor on the neuron output. Corresponds to the relative
+        amplitude of the output spikes of the neuron.
+
+    References
+    ----------
+    .. [1] Voelker, A. R., Rasmussen, D., & Eliasmith, C. (2020). A Spike in
+       Performance: Training Hybrid-Spiking Neural Networks with Quantized Activation
+       Functions. arXiv preprint arXiv:2002.03553. (https://arxiv.org/abs/2002.03553)
+    """
+
+    def __init__(self, base_type, amplitude=1.0):
+        super().__init__(base_type, amplitude=amplitude)
+
+        if "voltage" in base_type.probeable:
+            raise ValidationError(
+                "Cannot already include `voltage` in `probable`",
+                attr="base_type",
+                obj=self,
+            )
+
+    @property
+    def probeable(self):
+        base = list(self.base_type.probeable)
+        base[base.index("rates")] = "spikes"
+        base.insert(1 if base[0] == "spikes" else 0, "voltage")
+        return tuple(base)
+
+    def step_math(self, dt, J, output, voltage, *base_states):
```

Comment:
Should we use a double asterisk here, because in `SimNeurons.make_step` we always pass states by name (i.e. with a double asterisk)?